rust/libs/kvs/src/lib.rs

```diff
@@ -14,301 +14,690 @@
+            let val_bytes = codec.encode(&value)?;
+            let encoded_payload =
+                bincode::encode_to_vec((val_bytes.clone(), timestamp), bincode_standard())
+                    .map_err(|e| BidiError::Codec {
+                        source: Box::new(e),
+                    })?;
+            let encoded_inverse_payload =
+                bincode::encode_to_vec((key_bytes.clone(), timestamp), bincode_standard())
+                    .map_err(|e| BidiError::Codec {
+                        source: Box::new(e),
+                    })?;
+
+            let transaction_result = (&uo, &ou).transaction(move |(uo_tx, ou_tx)| {
+                let is_new;
+                if let Some(old_payload_ivec) = uo_tx.get(key_bytes.as_slice())? {
+                    is_new = false;
+                    let (old_val_bytes, _): (Vec<u8>, u128) =
+                        bincode::decode_from_slice(&old_payload_ivec, bincode_standard())
+                            .map(|(decoded, _)| decoded)
+                            .map_err(|e| {
+                                ConflictableTransactionError::Abort(BidiError::Codec {
+                                    source: Box::new(e),
+                                })
+                            })?;
+                    ou_tx.remove(old_val_bytes.as_slice())?;
+                } else {
+                    is_new = true;
+                }
+
+                uo_tx.insert(key_bytes.as_slice(), IVec::from(encoded_payload.clone()))?;
+                ou_tx.insert(
+                    val_bytes.as_slice(),
+                    IVec::from(encoded_inverse_payload.clone()),
+                )?;
+
+                Ok(is_new)
+            });
+
+            transaction_result.map_err(|e: TransactionError<BidiError>| {
+                BidiError::SledTransaction {
+                    source: Box::new(e),
+                }
+            })
+        })
+        .await??;
+
+        if was_inserted {
+            self.len.fetch_add(1, Ordering::SeqCst);
         }
+
         Ok(())
     }
 
-    fn len(&self) -> Result<usize> {
-        let ou_len = self.ou.len();
-        let uo_len = self.uo.len();
-        ensure!(ou_len == uo_len, "mismatch length, ou:{}, uo:{}", ou_len, uo_len);
-        Ok(ou_len)
+    /// Deletes a pair by its key and returns the associated value.
+    #[instrument(skip(self, key))]
+    pub async fn delete<Q>(&self, key: &Q) -> Result<V, BidiError>
+    where
+        K: Borrow<Q>,
+        Q: Serialize + Encode + ?Sized,
+    {
+        self.perform_delete(key, &self.uo, &self.ou).await
     }
 
-    fn close(&self) -> Result<()> {
-        Ok(())
+    /// Deletes a pair by its value and returns the associated key.
+    #[instrument(skip(self, value))]
+    pub async fn delete_inverse<Q>(&self, value: &Q) -> Result<K, BidiError>
+    where
+        V: Borrow<Q>,
+        Q: Serialize + Encode + ?Sized,
+    {
+        self.perform_delete(value, &self.ou, &self.uo).await
```

Comment:
`perform_delete` performs transactions in the order of inputs. Can this inverse order lead to deadlock?

Reply:
In this case, no deadlock will occur.

The reason is that sled's transaction API is designed to prevent this scenario. It automatically sorts all trees involved in a transaction by their names (in lexicographical order) before acquiring locks. This ensures a consistent global lock order, regardless of the order you pass the Trees in the tuple.

delete call: (&self.uo, &self.ou) -> sled sorts to ("ou", "uo") -> locks ou, then uo.

delete_inverse call: (&self.ou, &self.uo) -> sled sorts to ("ou", "uo") -> locks ou, then uo.

Since the lock acquisition order is always the same, a circular wait condition is impossible, thus preventing a deadlock.